intermediate_source/named_tensor_tutorial.py

```diff
@@ -0,0 +1,519 @@
+                  .align_to(..., 'H', 'T', 'D_head'))
+
+######################################################################
+# The first thing to note is how the code clearly states the input and
+# output dimensions: the input tensor must end with the ``T`` and ``D`` dims
+# and the output tensor ends in ``H``, ``T``, and ``D_head`` dims.
+#
+# The second thing to note is how clearly the code describes what is going on.
+# prepare_head takes the key, query, and value and splits the embedding dim into
+# multiple heads, finally rearranging the dim order to be ``[..., 'H', 'T', 'D_head']``.
+# ParlAI implements ``prepare_head`` as the following, using ``view`` and ``transpose``
+# operations:
+#
+# **(III) Explicit broadcasting by names**
+
+def prepare_head(tensor):
+    # input is [batch_size, seq_len, n_heads * dim_per_head]
+    # output is [batch_size * n_heads, seq_len, dim_per_head]
+    batch_size, seq_len, _ = tensor.size()
+    tensor = tensor.view(batch_size, tensor.size(1), n_heads, dim_per_head)
+    tensor = (
+        tensor.transpose(1, 2)
+        .contiguous()
+        .view(batch_size * n_heads, seq_len, dim_per_head)
+    )
+    return tensor
+
+######################################################################
+# Our named tensor variant uses ops that, though more verbose, also have
+# more semantic meaning than ``view`` and ``transpose`` and include enforcable
+# documentation in the form of names.
+#
+# **(III) Explicit broadcasting by names**
+
+def ignore():
+    # (III)
+    attn_mask = (attn_mask == 0).align_as(dot_prod)
+
+    dot_prod.masked_fill_(attn_mask, -float(1e20))
+
+######################################################################
+# ``mask`` usually has dims ``[N, T]`` (in the case of self attention) or
+# ``[N, T, T_key]`` (in the case of encoder attention) while ``dot_prod``
+# has dims ``[N, H, T, T_key]``. To make ``mask`` broadcast correctly with
+# ``dot_prod``, we would usually `unsqueeze` dims ``1`` and ``-1`` in the case of self
+# attention or ``unsqueeze`` dim ``1`` in the case of encoder attention. Using
+# named tensors, we simply align ``attn_mask`` to ``dot_prod`` using ``align_as``
+# and stop worrying about where to ``unsqueeze`` dims.
+#
+# **(IV) More dimension manipulation using ``align_to`` and ``flatten``**
+
+def ignore():
+    # (IV)
+    attentioned = (
+        attn_weights.matmul(v).refine_names(..., 'H', 'T', 'D_head')
+        .align_to(..., 'T', 'H', 'D_head')
+        .flatten(['H', 'D_head'], 'D')
+    )
+
+######################################################################
+# (IV): Like (II), ``align_to`` and ``flatten`` are more semantically
+# meaningful than `view` (despite being more verbose).
+#
+# Running the example
+# -------------------
+
+n, t, d, h = 7, 5, 2 * 3, 3
+query = torch.randn(n, t, d, names=('N', 'T', 'D'))
+mask = torch.ones(n, t, names=('N', 'T'))
+attn = MultiHeadAttention(h, d)
+output = attn(query, mask=mask)
+print(output.names)
+

```

Comment:
We're missing the last section from the updated tutorial I posted: 
```
Notice how in the code we did not mention the name of the batch dimension at all. In fact, the code is agnostic to the existence of batch dimensions; we can run the following example-level code:

query = torch.randn(t, d, names=('T', 'D'))
mask = torch.ones(t, names=('T',))
output = attn(query, mask=mask)
output.names
('T', 'D')
```
(https://our.internmc.facebook.com/intern/anp/view/?id=152205 scroll to the end)